feat($resource): add support for timeout in cancellable actions

mrac opened the pull request:
Old behavior: actions can be either cancellable or have a numeric timeout.
When having both defined, cancellable was ignored.
With this commit: it's possible for actions to have both cancellable:true
and numeric timeout defined.

Example usage:

``` js
var Post = $resource('/posts/:id', {id: '@id'}, {
  get: {
    method: 'GET',
    cancellable: true,
    timeout: 10000
  }
});

var currentPost = Post.get({id: 1});
...
// the next request can cancel the previous one
currentPost.$cancelRequest();
currentPost = Post.get({id: 2});

// any of those requests will also timeout, if the response
// doesn't come within 10 seconds
```

mrac committed:
feat($resource): add support for timeout in cancellable actions

Old behavior: actions can be either cancellable or have a numeric timeout.
When having both defined, cancellable was ignored.
With this commit: it's possible for actions to have both cancellable:true
and numeric timeout defined.

Example usage:

```js
var Post = $resource('/posts/:id', {id: '@id'}, {
  get: {
    method: 'GET',
    cancellable: true,
    timeout: 10000
  }
});

var currentPost = Post.get({id: 1});
...
// the next request can cancel the previous one
currentPost.$cancelRequest();
currentPost = Post.get({id: 2});

// any of those requests will also timeout, if the response
// doesn't come within 10 seconds
```

gkalpak commented on src/ngResource/resource.js:
```diff
@@ -586,9 +590,11 @@ angular.module('ngResource', ['ng']).
+          } else {
+            if(cancellable) {
+              simulatedTimeout = action.timeout;
+              delete action.timeout;
```
Why delete it here ? Couldn't we just let it be ?

mrac replied:
Yes we don't need to delete it. It gets overwritten anyway later:
```
httpConfig.timeout = timeoutDeferred.promise;
```

gkalpak replied:
I don't think it is overwritten. `action.timeout` is not the same as `httpConfig.timeout`.
But we shouldn't delete it if we don't have to :smiley:

gkalpak commented on src/ngResource/resource.js:
```diff
@@ -707,6 +722,7 @@ angular.module('ngResource', ['ng']).
               if (!isInstanceCall && cancellable) {
                 value.$cancelRequest = angular.noop;
                 timeoutDeferred = httpConfig.timeout = null;
+                $timeout.cancel(simulatedTimeoutPromise);
```
Also set `simulatedTimeoutPromise` to null (just in case).

mrac replied:
Ok.

gkalpak commented on src/ngResource/resource.js:
```diff
@@ -575,7 +575,11 @@ angular.module('ngResource', ['ng']).
+            (options && angular.isDefined(options.cancellable)) ? options.cancellable :
+              provider.defaults.cancellable;
+
+          var simulatedTimeout;
```
`simulatedTimeout` is a little confusing imo (what is "simulated" about it).
<sub>(On the other hand, I'm notiriously bad with names :smiley:)</sub>

mrac replied:
I was trying to find a different name, not to confuse with timeout which is passed to $http. There are quite a few variables that deal with $http timeout.

So it for me it means "non-$http" timeout.

gkalpak commented on src/ngResource/resource.js:
```diff
@@ -656,6 +663,14 @@ angular.module('ngResource', ['ng']).
+                  if(timeoutDeferred) {
+                    timeoutDeferred.resolve();
+                  }
+                }, simulatedTimeout);
```
This could be just `simulatedTimeoutPromise = $timeout(timeoutDeferred.resolve, simulatedTimeout);`.

mrac replied:
Yes, but I wanted to be sure, timeoutDeferred is not null already, so extra if() here. But may be it's obvious.

mrac replied:
You are right, timeoutDeferred.resolve can be evaluated instantly (and not inside the timeout handler).

gkalpak commented:
A couple of minor comments, but LGTM otherwise.

mrac committed:
fixup: simplify $timeout handler

simplify $timeout handler
set `simulatedTimeoutPromise` to null
don't delete action.timeout - not necessary

gkalpak commented:
@mrac, thx for working on this :sparkles: :+1: :sparkles: 
I will take it from here (due to tight time constraints to catch the 1.5.0 train).

BTW (so you know for the next time), there were some legitimate failures in tests - nothing major just a couple of whitespace issues that JSCS (rightfully) complaints about :smiley:
(You can see them on Travis, e.g. [here][1] - or you can also run the tests locally with `grunt test` (to run all the tests) or `grunt ci-checks`/`grunt test:unit`/`grunt test:e2e` etc (to run a specific subset).)

[1]: https://travis-ci.org/angular/angular.js/jobs/104696540

mrac commented:
@gkalpak great, thanks! Good to hear it's going to be in 1.5.0.
Yeah, I ran unit tests and e2e tests as described here https://docs.angularjs.org/misc/contribute. Apparently ```grunt test``` does much more ;) Thanks for the info.

mrac commented:
@gkalpak It's nice to have it released already, thanks!

Still here remains a problem of how to distinguish between cancelled requests and timed-out requests. The latter might need different handling in most cases. Apparently both produce the same error response `{status: -1, statusText: "", ...}`

The solution would be to allow to pass a statusConfig to timeout-promise resolve value in $http. It could produce different error status and statusText. Currently timeout-promise resolve value is ignored entirely.

gkalpak commented:
I don't think it's high priority, but I would be happy to review any PRs if anyone decided to submit them :wink:

mrac commented:
@gkalpak Great! So take a look:
feat($http,$resource): statusText resolved in timeout promise #14021
